src/repositories/updater.rs

```diff
@@ -0,0 +1,354 @@
+
+    /// How many items we can query in one graphql request.
+    fn chunk_size(&self) -> usize;
+
+    /// Used by both backfill_repositories and load_repository. When the repository is missing
+    /// `None` is returned.
+    fn fetch_repository(&self, name: &RepositoryName) -> Result<Option<Repository>>;
+
+    /// Used by update_all_crates.
+    ///
+    /// The returned struct will contain all the information needed for `RepositoriesUpdater` to
+    /// update repositories that are still present and delete the missing ones.
+    fn fetch_repositories(&self, ids: &[String]) -> Result<FetchRepositoriesResult>;
+}
+
+pub struct Repository {
+    pub id: String,
+    pub name_with_owner: String,
+    pub description: Option<String>,
+    pub last_activity_at: Option<DateTime<Utc>>,
+    pub stars: i64,
+    pub forks: i64,
+    pub issues: i64,
+}
+
+#[derive(Default)]
+pub struct FetchRepositoriesResult {
+    pub present: HashMap<String, Repository>,
+    pub missing: Vec<String>,
+}
+
+pub struct RepositoryStatsUpdater {
+    updaters: Vec<Box<dyn RepositoryForge + Send + Sync>>,
+}
+
+impl fmt::Debug for RepositoryStatsUpdater {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "RepositoryStatsUpdater {{ {:?} }}",
+            self.updaters.iter().map(|u| u.host()).collect::<Vec<_>>()
+        )
+    }
+}
+
+impl RepositoryStatsUpdater {
+    pub fn new(config: &Config) -> Self {
+        let mut updaters: Vec<Box<dyn RepositoryForge + Send + Sync>> = Vec::with_capacity(3);
+        if let Ok(Some(updater)) = GitHub::new(&config) {
+            updaters.push(Box::new(updater));
+        }
+        if let Ok(Some(updater)) = GitLab::new("gitlab.com", &config.gitlab_accesstoken) {
+            updaters.push(Box::new(updater));
+        }
+        if let Ok(Some(updater)) = GitLab::new("gitlab.freedesktop.org", &None) {
+            updaters.push(Box::new(updater));
+        }
+        Self { updaters }
+    }
+
+    pub(crate) fn load_repository(
+        &self,
+        conn: &mut Client,
+        metadata: &MetadataPackage,
+    ) -> Result<Option<i32>> {
+        let url = match &metadata.repository {
+            Some(url) => url,
+            None => {
+                debug!("did not collect stats as no repository URL was present");
+                return Ok(None);
+            }
+        };
+        self.load_repository_inner(conn, url)
+    }
+
+    fn load_repository_inner(&self, conn: &mut Client, url: &str) -> Result<Option<i32>> {
+        let name = match repository_name(url) {
+            Some(name) => name,
+            None => return Ok(None),
+        };
+
+        // Avoid querying the APIs for repositories we already loaded.
+        if let Some(row) = conn.query_opt(
+            "SELECT id FROM repositories WHERE name = $1 AND host = $2 LIMIT 1;",
+            &[&format!("{}/{}", name.owner, name.repo), &name.host],
+        )? {
```

Comment:
This is wrong for GitLab, as GitLab's `nameWithNamespace` does not match the owner and repo in the URL.

For example, https://gitlab.com/crates.rs/crates.rs has a `nameWithNamespace` of `lib.rs / Lib.rs 🔰 main project`. In order for this to work we'd need to replace `nameWithNamespace` on the GraphQL query with `fullPath`.

Reply:
Currently, `fullPath` is that as you can see in `GitLab::fetch_repository`. So if this is an issue it'll require a bigger update. But I'm still not sure about what it's supposed to look like.

Reply:
Sorry, misunderstood your comment. Easy fix in fact...